FlowCal/plot.py

```diff
@@ -64,6 +67,510 @@
+    taken from the largest ``data[i].range(channel)[1]`` or the largest
+    element in ``data[i]`` if ``data[i].range()`` is not available, M is
+    set to ``4.5 / np.log10(262144) * np.log10(T)``, and W is taken from
+    ``(M - log10(T / abs(r))) / 2``, where ``r`` is the minimum negative
+    event. If no negative events are present, W is set to zero.
+
+    References
+    ----------
+    .. [1] D.R. Parks, M. Roederer, W.A. Moore, "A New Logicle Display
+    Method Avoids Deceptive Effects of Logarithmic Scaling for Low Signals
+    and Compensated Data," Cytometry Part A 69A:541-551, 2006, PMID
+    16604519.
+
+    """
+    # ``input_dims``, ``output_dims``, and ``is_separable`` are required by
+    # matplotlib.
+    input_dims = 1
+    output_dims = 1
+    is_separable = True
+    # Locator objects need this object to store the logarithm base used as an
+    # attribute.
+    base = 10
+
+    def __init__(self, T=None, M=None, W=None, data=None, channel=None):
+        matplotlib.transforms.Transform.__init__(self)
+        # If data is included, try to obtain T, M and W from it
+        if data is not None:
+            if channel is None:
+                ValueError("if data is provided, a channel should be specified")
+            # Convert to list if necessary
+            if not isinstance(data, list):
+                data = [data]
+            # Obtain T, M, and W if not specified
+            # If elements of data have ``.range()``, use it to determine the
+            # max data value. Else, use the maximum value in the array.
+            if T is None:
+                T = 0
+                for d in data:
+                    # Extract channel
+                    y = d[:, channel] if d.ndim > 1 else d
+                    if hasattr(y, 'range') and hasattr(y.range, '__call__'):
+                        Ti = y.range(0)[1]
+                    else:
+                        Ti = np.max(y)
+                    T = Ti if Ti > T else T
+            if M is None:
+                M = 4.5 / np.log10(262144) * np.log10(T)
+            if W is None:
+                W = 0
+                for d in data:
+                    # Extract channel
+                    y = d[:, channel] if d.ndim > 1 else d
+                    # If negative events are present, use minimum.
+                    if np.any(y < 0):
+                        r = np.min(y)
+                        Wi = (M - np.log10(T / abs(r))) / 2
+                        W = Wi if Wi > W else W
+        else:
+            # Default parameter values
+            if T is None:
+                T = 262144
+            if M is None:
+                M = 4.5
+            if W is None:
+                W = 0.5
+        # Check that property values are valid
+        if T <= 0:
+            ValueError("T should be positive")
+        if M <= 0:
+            ValueError("M should be positive")
+        if W < 0:
+            ValueError("W should not be negative")
```

Comment:
should these errors be `raise`ed? or is this some black magic that i'm not familiar with? Same for the `ValueError` up at the beginning of `__init__()`